plugins/inputs/aliyuncms/aliyuncms.go

```diff
@@ -0,0 +1,597 @@
+	  #dimensions = '{"instanceId": "p-example"}'
+	
+	  ## Or you can specify several dimensions at once:
+	  #dimensions = '[{"instanceId": "p-example"},{"instanceId": "q-example"}]'
+	
+	  ## Enrichment tags, can be added from discovery (if supported)
+	  ## Notation is <measurement_tag_name>:<JMES query path (https://jmespath.org/tutorial.html)>
+	  ## To figure out which fields are available, consult the Describe<ObjectType> API per project.
+	  ## For example, for SLB: https://api.aliyun.com/#/?product=Slb&version=2014-05-15&api=DescribeLoadBalancers&params={}&tab=MOCK&lang=GO
+	  #tag_query_path = [
+	  #    "address:Address",
+	  #    "name:LoadBalancerName",
+	  #    "cluster_owner:Tags.Tag[?TagKey=='cs.cluster.name'].TagValue | [0]"
+	  #    ]
+	  ## The following tags added by default: regionId (if discovery enabled), userId, instanceId.
+	
+	  ## Allow metrics without discovery data, if discovery is enabled. If set to true, then metric without discovery
+	  ## data would be emitted, otherwise dropped. This cane be of help, in case debugging dimension filters, or partial coverage 
+	  ## of discovery scope vs monitoring scope 
+	  #allow_dps_without_discovery = false
+`
+	inputTitle = "inputs.aliyuncms"
+	inputName  = "aliyuncms"
+)
+
+type logger struct {
+	errHeader    string
+	warnHeader   string
+	infoHeader   string
+	debugHeader  string
+	headerFormat string
+}
+
+var lg = newLogger(inputTitle)
+
+func newLogger(header string) *logger {
+	return &logger{
+		headerFormat: "%s %s",
+		errHeader:    fmt.Sprintf("E! [%s]", header),
+		warnHeader:   fmt.Sprintf("W! [%s]", header),
+		infoHeader:   fmt.Sprintf("I! [%s]", header),
+		debugHeader:  fmt.Sprintf("D! [%s]", header)}
+}
+func (l logger) logE(format string, v ...interface{}) {
+	log.Printf(l.headerFormat, l.errHeader, fmt.Sprintf(format, v...))
+}
+func (l logger) logW(format string, v ...interface{}) {
+	log.Printf(l.headerFormat, l.warnHeader, fmt.Sprintf(format, v...))
+}
+func (l logger) logI(format string, v ...interface{}) {
+
+	log.Printf(l.headerFormat, l.infoHeader, fmt.Sprintf(format, v...))
+}
+func (l logger) logD(format string, v ...interface{}) {
+	log.Printf(l.headerFormat, l.debugHeader, fmt.Sprintf(format, v...))
+}
+
+type (
+	// AliyunCMS is aliyun cms config info.
+	AliyunCMS struct {
+		AccessKeyID       string `toml:"access_key_id"`
+		AccessKeySecret   string `toml:"access_key_secret"`
+		AccessKeyStsToken string `toml:"access_key_sts_token"`
+		RoleArn           string `toml:"role_arn"`
+		RoleSessionName   string `toml:"role_session_name"`
+		PrivateKey        string `toml:"private_key"`
+		PublicKeyID       string `toml:"public_key_id"`
+		RoleName          string `toml:"role_name"`
+
+		DiscoveryRegions  []string          `toml:"discovery_regions"`
+		DiscoveryInterval internal.Duration `toml:"discovery_interval"`
+		Period            internal.Duration `toml:"period"`
+		Delay             internal.Duration `toml:"delay"`
```

Comment:
is it possible to take out the aggregation parts? usually we use separate plugins for that. Is there something special about how this does aggregation?

Reply:
Yes, there is some special logic here. The API supports different aggregation per object, we need to clearly specify which aggregation period we are querying (that's the reason for `period` parameter). The logic is similar to cloudwatch input plugin, since Alicloud borrows a lot of concepts from AWS.